### ISSUE TITLE
Can't get 88% :(

### ISSUE THREAD

**Issue by chledowski** [adamzjk/SNLI-Keras#3]
Hello Adam!

I am trying to reproduce 88% on ESIM by using your nice repo, but I am able to obtain about 87%. You wrote that your model has 4.6m parameters, while mine has 4.3m. Could you share your model summary with me, to help me see what I am lacking?

```__________________________________________________________________________________________________                                                                                                          
Layer (type)                    Output Shape         Param #     Connected to                                                                                                                               
==================================================================================================                                                                                                          
input_1 (InputLayer)            (None, 42)           0                                                                                                                                                      
__________________________________________________________________________________________________                                                                                                          
input_2 (InputLayer)            (None, 42)           0                                                                                                                                                      
__________________________________________________________________________________________________                                                                                                          
embedding_1 (Embedding)         (None, 42, 300)      12717600    input_1[0][0]                                                                                                                              
                                                                 input_2[0][0]                                                                                                                              
__________________________________________________________________________________________________                                                                                                          
bidirectional_1 (Bidirectional) (None, 42, 600)      1444800     embedding_1[0][0]                                                                                                                          
                                                                 embedding_1[1][0]                                                                                                                          
__________________________________________________________________________________________________                                                                                                          
dropout_1 (Dropout)             (None, 42, 600)      0           bidirectional_1[0][0]                                                                                                                      
__________________________________________________________________________________________________                                                                                                          
dropout_2 (Dropout)             (None, 42, 600)      0           bidirectional_1[1][0]                                                                                                                      
__________________________________________________________________________________________________                                                                                                          
dot_1 (Dot)                     (None, 42, 42)       0           dropout_2[0][0]                                                                                                                            
                                                                 dropout_1[0][0]                                                                                                                            
__________________________________________________________________________________________________                                                                                                          
permute_1 (Permute)             (None, 42, 42)       0           dot_1[0][0]                                                                                                                                
__________________________________________________________________________________________________                                                                                                          
lambda_2 (Lambda)               (None, 42, 42)       0           permute_1[0][0]                                                                                                                            
__________________________________________________________________________________________________                                                                                                          
lambda_1 (Lambda)               (None, 42, 42)       0           dot_1[0][0]
__________________________________________________________________________________________________
dot_2 (Dot)                     (None, 42, 600)      0           lambda_2[0][0]
                                                                 dropout_2[0][0]
__________________________________________________________________________________________________
dot_3 (Dot)                     (None, 42, 600)      0           lambda_1[0][0]
                                                                 dropout_1[0][0]
__________________________________________________________________________________________________
subtract_1 (Subtract)           (None, 42, 600)      0           dropout_1[0][0]
                                                                 dot_2[0][0]
__________________________________________________________________________________________________
multiply_1 (Multiply)           (None, 42, 600)      0           dropout_1[0][0]
                                                                 dot_2[0][0]
__________________________________________________________________________________________________
subtract_2 (Subtract)           (None, 42, 600)      0           dropout_2[0][0]
                                                                 dot_3[0][0]
__________________________________________________________________________________________________                                                                                                          
multiply_2 (Multiply)           (None, 42, 600)      0           dropout_2[0][0]                                                                                                                            
                                                                 dot_3[0][0]                                                                                                                                
__________________________________________________________________________________________________                                                                                                          
concatenate_1 (Concatenate)     (None, 42, 2400)     0           dropout_1[0][0]                                                                                                                            
                                                                 dot_2[0][0]                                                                                                                                
                                                                 subtract_1[0][0]                                                                                                                           
                                                                 multiply_1[0][0]                                                                                                                           
__________________________________________________________________________________________________
concatenate_2 (Concatenate)     (None, 42, 2400)     0           dropout_2[0][0]
                                                                 dot_3[0][0]
                                                                 subtract_2[0][0]
                                                                 multiply_2[0][0]
__________________________________________________________________________________________________
dropout_3 (Dropout)             (None, 42, 2400)     0           concatenate_1[0][0]
__________________________________________________________________________________________________
dropout_4 (Dropout)             (None, 42, 2400)     0           concatenate_2[0][0]
__________________________________________________________________________________________________
translate (TimeDistributed)     (None, 42, 300)      720300      dropout_3[0][0]
                                                                 dropout_4[0][0]
__________________________________________________________________________________________________
finaldecoder (Bidirectional)    (None, 42, 600)      1444800     translate[0][0]
                                                                 translate[1][0]
__________________________________________________________________________________________________
dropout_5 (Dropout)             (None, 42, 600)      0           finaldecoder[0][0]
__________________________________________________________________________________________________
dropout_6 (Dropout)             (None, 42, 600)      0           finaldecoder[1][0]
__________________________________________________________________________________________________
lambda_3 (Lambda)               (None, 600)          0           dropout_5[0][0]
                                                                 dropout_6[0][0]
__________________________________________________________________________________________________
lambda_4 (Lambda)               (None, 600)          0           dropout_5[0][0]
                                                                 dropout_6[0][0]
__________________________________________________________________________________________________
concatenate_3 (Concatenate)     (None, 2400)         0           lambda_3[0][0]
                                                                 lambda_4[0][0]
                                                                 lambda_3[1][0]
                                                                 lambda_4[1][0]
__________________________________________________________________________________________________
dropout_7 (Dropout)             (None, 2400)         0           concatenate_3[0][0]
__________________________________________________________________________________________________
dense300_snli (Dense)           (None, 300)          720300      dropout_7[0][0]
__________________________________________________________________________________________________
activation_1 (Activation)       (None, 300)          0           dense300_snli[0][0]
__________________________________________________________________________________________________
dropout_8 (Dropout)             (None, 300)          0           activation_1[0][0]
__________________________________________________________________________________________________
judge300_snli (Dense)           (None, 3)            903         dropout_8[0][0]
==================================================================================================
Total params: 17,048,703
Trainable params: 4,331,103
Non-trainable params: 12,717,600
__________________________________________________________________________________________________
```
This is how my model summary looks like. I am also experimenting with training the embeddings and using CuDNNLSTM layer to speed up the training.

**Comment by naayaaa** [comment]
In fact i can't get 88% too, the 88% accuracy reported in the paper is produced by an ensemble of multiple models, so 87% of a single model is already a pretty good result.

**Comment by chledowski** [comment]
Hi Adam,

We have managed to get 88% without ensembling. We have changed initialization and our preprocessing to match the original paper, and it worked.

Anyway, thanks for your code, it helped a lot! :+1:

**Comment by N-Almarwani** [comment]
Hi @chledowski
Would you please specify which initilization did you use and which layer did you change in this model

**Comment by chledowski** [comment]
Hi,

we copied the preprocessing from the original papers code. I think it helped us a little. It was in Java if I remember correctly. We need to rewrite it to python eventually.. ;) 

Apart from that we tried many things. We noticed that there are discrepancies between this and the original code. If I remember correctly, dropout was in wrong places, layers like bilstm had different initializers.

We were using all parameters as in the original code. That means bs 32. With bs 512 we get about 87.5 acc. 

If you have more doubts don't hesitate to ask. :)
